### PR TITLE
SWORD1: Implement palette fades

### PR DESCRIPTION
**Pull request by AndywinXp**
Ahoy! This is the first one of a series of changes and PRs which aim at finishing and implementing all those things from the original Broken Sword 1 executable which we are currently missing in our implementation. We have a great port of the engine and it would be a shame not to finish those last tasks; after all the devil is in the details 🙂

A full list can be seen here: https://wiki.scummvm.org/index.php?title=Sword1

## What am I looking at here?

In this first part I have implemented accurate palette fade-ins and fade-outs. In order to achieve this, one of the necessities was to fix the game timers, which was part of the to-do plan anyway.
Also, I implemented fast and slow modes from the original debug commands because at one point I was going insane debugging this 🙂 

All of this was implemented from the original code, using the DOS implementation as the reference (hence the usage of some kind of emulation of an vertical blank interrupt). Of course this is not the DOS code verbatim:
* Since we apparently don't have files for the WIN95 drivers but only for the DOS ones, I manually rewrote the palette fading routine from x86 asm to C++;
* The original DOS version uses the vblank timer for the top and bottom menus animations, palette fades and in-game frame limiter. In this implementation, the latter is not using the vblank simulation but instead uses our current timer system (with the correct frame time though). This is because our current system is more precise and smooth than an emulated vblank interrupt system (I've reluctantly tried and I'm glad that it failed; I don't want a threaded game timer 🙂).
* The code for fades in the main menu seems to work like it should, but it's adapted for our control panel implementation which is quite different from the original. Said implementation will be changed to match the source code in one of the next PRs for this project (because of the big amount of differences noted in the wiki link above). Needless to say, this future reimplementation will take into account all past contributions and changes (by our past and current devs) which don't have a match on the source code in our possession, for e.g. Mac and PSX support, etc.

Let me know how is it. The moment this is merged, work will be resumed on all the other tasks mentioned on the wiki. Alternatively, if you prefer a bigger PR with more stuff poured into it, let me know.

### CODE REVIEW

**Comment by bluegr** [comment]
Nice work!

It would be better to keep PRs as targeted as possible, so it would be preferable to create separate PRs for the other TODOs

**Comment by AndywinXp** [comment]
Thanks! I'll do as suggested then 🙂

File: engines/sword1/mouse.cpp
Comment (sev-) [inline engines/sword1/mouse.cpp:74]: Is this an intended change?

File: engines/sword1/screen.cpp
Comment (sev-) [inline engines/sword1/screen.cpp:230]: Simplify the indentation?

```C++
if (_paletteFadeInfo.fadeCount != _paletteFadeInfo.paletteIndex)
  return;

... rest of the method
```

File: engines/sword1/screen.cpp
Comment (sev-) [inline engines/sword1/screen.cpp:258]: Perhaps, I am too picky, but I'd write something like:

```C++
for (int i = 0; i < 256 * 3; i++) {
   if (_paletteFadeInfo.srcPalette[i] > 0 && (int8)_paletteFadeInfo.srcPalette[i] < (int8)_paletteFadeInfo.dstPalette[i])
    _paletteFadeInfo.srcPalette[i] += _paletteFadeInfo.paletteStatus;
 ```

Since `FADE_UP = 1` and `FADE_DOWN = -1`

File: engines/sword1/screen.h
Comment (sev-) [inline engines/sword1/screen.h:111]: 32 or 31? E.g. one bit set or 5, following 63

File: engines/sword1/sword1.cpp
Comment (sev-) [inline engines/sword1/sword1.cpp:790]: Is 2 seconds enough? What is the maximum fade duration?

File: engines/sword1/sword1.cpp
Comment (sev-) [inline engines/sword1/sword1.cpp:870]: I'd suggest writing `TODO: Fade audio here`, the same below.

**Review by sev- — CHANGES_REQUESTED** [review]
Good stuff! Have some remarks/questions

File: engines/sword1/mouse.cpp
Comment (AndywinXp) [reply engines/sword1/mouse.cpp:74]: Yes, that's correct 🙂 the luggage operation is to be done by calling the logic function `fnSetLuggage` at the beginning of the main menu and not the `setLuggage` operation in this part. Moreover the game _must not restore the luggage_ (which is basically the action cursor), hence the other change just below this one.

File: engines/sword1/screen.cpp
Comment (AndywinXp) [reply engines/sword1/screen.cpp:230]: Sure!

File: engines/sword1/screen.h
Comment (AndywinXp) [reply engines/sword1/screen.h:111]: Just re-checked the source code for good measure: it actually uses 32 (and 63 above) 🙂

File: engines/sword1/sword1.cpp
Comment (AndywinXp) [reply engines/sword1/sword1.cpp:870]: Sure!

File: engines/sword1/sword1.cpp
Comment (AndywinXp) [reply engines/sword1/sword1.cpp:790]: Yes; basically the fade happens at a 60Hz rate within the simulated interrupt. Every fade operates for 64 steps, so this means: `16.67 ms/step * 64 steps = 1066.88 ms`. When performing some debugging operation, the timer thread can surely have some hiccups, so I took it to 2 seconds just to be safe.

File: engines/sword1/screen.cpp
Comment (AndywinXp) [reply engines/sword1/screen.cpp:258]: Your way of writing this is certainly more concise and shorter; unfortunately I have tested it and it seems to break fades 🙁 It appears the two conditions can't be used together. 

If you are okay with it I'd leave the verbose code on

Comment (lephilousophe) [reply engines/sword1/screen.cpp]: This indeed can't work as when fading up from black, srcPalette will be 0 and fading will never happen.
Maybe CLIP could be used but it seems safer to keep it the way it's written here as there could be some wrapping (if we fade from white to black for example).

File: engines/sword1/screen.cpp
Comment (criezy) [inline engines/sword1/screen.cpp:1342]: Do you know where in the game this function is called?

The Mac version of the game is known to use color 255 instead of 0 in some places, for example for the transparent color in parallax, or the background color visible in some screens, such as when opening the door in Nico's apartment.

**Review by criezy — CHANGES_REQUESTED** [review]
> The Mac version of the game is known to use color 255 instead of 0 in some places, for example for the transparent color in parallax, or the background color visible in some screens, such as when opening the door in Nico's apartment.

And that recollection prompted me to test those places and find out that this pull request is actually breaking that and introducing a regression. When opening the door the color is now red instead of black:
image

By the way, while it seems to work I think the current code for the Mac version is probably not fully correct. I should take a look at the disassembly to check exactly what the original does.

File: engines/sword1/screen.cpp
Comment (AndywinXp) [reply engines/sword1/screen.cpp:1342]: This function was previously unimplemented as it was called (mostly) in debug opcodes (and that comes with the intention of implementing debug stuff in the future). I said "mostly" because there's an actual occurence of `FN_blue_flash()` in what seems to be a define macro for the PSX version of `FN_walk` which I haven't yet checked if we have in our system or if it's actually needed.

I can definitely check a Mac disassembly tonight and let you know.

File: engines/sword1/screen.cpp
Comment (AndywinXp) [reply engines/sword1/screen.cpp:1347]: Sure, you're absolutely right!

File: engines/sword1/control.cpp
Comment (AndywinXp) [reply engines/sword1/control.cpp:353]: This is done to prevent a read access violation within the PSX version. As mentioned in the PR description, the control panel code is the next thing to be rewritten to match the source code (which appears to be the same with minor differences throughout PC, Mac and PSX), because right now it's very different and I had to hack it around to make fades work in there. 

This goes with a promise that there won't be such strange calls in the code.

File: engines/sword1/control.cpp
Comment (AndywinXp) [reply engines/sword1/control.cpp:354]: Yes, you are correct, done!

File: engines/sword1/screen.cpp
Comment (AndywinXp) [reply engines/sword1/screen.cpp:197]: Sure, I'll change both things accordingly!

**Comment by AndywinXp** [comment]
Hi @criezy, thanks for the review! About the Mac concerns: tonight I can definitely help and pull up a disassembly as well to try and see what is actually supposed to happen in the cases the screen has to fade from and to black.

File: engines/sword1/screen.cpp
Comment (AndywinXp) [reply engines/sword1/screen.cpp:1342]: About the Mac issue; I dug into the PPC disassembly and the palette handling seem to be identical to the PC version.

So if I patch `fnSetFadeTargetPalette()` to be like this (sorry, I have to reintroduce the const_cast 🤷‍♂️) it works correctly:

```c++
void Screen::fnSetFadeTargetPalette(uint8 start, uint16 length, uint32 id, bool toBlack) {
	uint8 *rgbData = nullptr;

	if (toBlack) {
		rgbData = const_cast<uint8 *>(_black);
	} else {
		rgbData = (uint8 *)_resMan->openFetchRes(id);
	}

	// Bug #8636: Force color 255 to black
	if (SwordEngine::isMac()) {
		if (start != 0 && start + length == 256)
			rgbData[(length - 1) * 3 + 0] = rgbData[(length - 1) * 3 + 1] = rgbData[(length - 1) * 3 + 2] = 0;
	}

	if (SwordEngine::_systemVars.wantFade) {
		memcpy(_currentPalette + (start * 3), rgbData, length * 3);
		memcpy(_paletteFadeInfo.dstPalette + (start * 3), rgbData, length * 3);
		memset(_paletteFadeInfo.srcPalette + (start * 3), 0, length * 3);
	} else {
		memcpy(_currentPalette + (start * 3), rgbData, length * 3);
	}

	_resMan->resClose(id);
}
```

The solution to this dilemma seems to be to apply your patch here as well, as you can see. If you approve it I will commit that.

Comment (criezy) [reply engines/sword1/screen.cpp:1342]: The test on `start != 0` is I think one of the things that does not make sense in the current code and just works by accident. I can't think of any good reason why I added that check here.

But in any case, I think that rather than add the hack here, it would make more sense to modify `Screen::updateScreen()`:
```diff
void Screen::updateScreen() {
	if (Logic::_scriptVars[NEW_PALETTE]) {
		_updatePalette = true;
		Logic::_scriptVars[NEW_PALETTE] = 0;
	}

	if (_updatePalette) {
		fnSetFadeTargetPalette(0, 184, _roomDefTable[_currentScreen].palettes[0]);
		fnSetFadeTargetPalette(184, 72, _roomDefTable[_currentScreen].palettes[1]);
		fnSetFadeTargetPalette(0, 1, 0, true);
+		// Bug #8636: Force color 255 to black
+		if (SwordEngine::isMac())
+			fnSetFadeTargetPalette(255, 1, 0, true);
		startFadePaletteUp(1);
		_updatePalette = false;
	}
```

Comment (AndywinXp) [reply engines/sword1/screen.cpp:1342]: Thanks, definitely cleaner, love it. Commited!

**Comment by criezy** [comment]
Savegame at the St Ninian's church: [sword1.023.zip](https://github.com/scummvm/scummvm/files/12479846/sword1.023.zip)

**Comment by AndywinXp** [comment]
This was a very nice catch, thanks! And this is one of those times in which I'm grateful to have the original scripts at hand...

What is basically happen is that this ending scene is not a Smacker video but is instead a concatenation of (as they call them) "screens" functioning as camera angles. Normally when a screen is changed, a palette fade happens. But in this case, right before the sequence begins...
```
FN_set_palette_to_cut();	// all 'cuts' during this sequence, rather than fades
```

And it works! Kind of! There are no fades but the screen is dim as soon as a screen/palette change happens. Why? Because I forgot to left-shift the palette when a fade-cut was signaled to happen 🙂

**Comment by AndywinXp** [comment]
> To answer my own question, the version from GOG has black bars there. So this appears to be a mac-specific issue (I must say I am not really surprised).

Do we have any video of someone playing this version on original hardware? I bet this could shed some light about whatever should happen in this case.

**Comment by criezy** [comment]
> Do we have any video of someone playing this version on original hardware? I bet this could shed some light about whatever should happen in this case.

Note that I know. But I have an old MacOS 9 laptop that still works so I can give it a go. I will probably have to play from the start as I doubt I still have savegames for the original. Not that I mind, but it will probably take a little while to reach that point which is unfortunately at the end of the game.

**Comment by criezy** [comment]
>  Because I forgot to left-shift the palette when a fade-cut was signaled to happen

Ah indeed. And that reminds of another "issue" that I forgot to mention because it is actually harmless. The colors that you have added to the `Screen` class (`_white`, `_red`, `_black`, ...) are already shifted (so using 8 bits). There is however one place where one of those colors is used as a non-shifted (6-bits) color: in `Screen::fnSetFadeTargetPalette`. This does not matter though as that color is `_black`, and shifted or not shifted a 0 is still a 0.

**Comment by AndywinXp** [comment]
> > Do we have any video of someone playing this version on original hardware? I bet this could shed some light about whatever should happen in this case.
> 
> Not that I know. But I have an old MacOS 9 laptop that still works so I can give it a go. I will probably have to play from the start as I doubt I still have savegames for the original. Not that I mind, but it will probably take a little while to reach that point which is unfortunately at the end of the game.

Hopefully this can help, launch the game from command line like this:
```
sword rev <boot-param>
```

where <boot-param> can be chosen from [these ones ](https://wiki.scummvm.org/index.php?title=Boot_Params#Broken_Sword_1:_The_Shadow_of_the_Templars). The main menu should appear if you already have savegames, in that case, just press restart. They do work on the dos executable.

**Review by criezy — APPROVED** [review]
Thank you. This looks good to me now.

The remaining issue with the mac version is not a regression from this pull request and already occurs with the current code from master. So there is no reason to block this pull request because of it.

**Comment by AndywinXp** [comment]
Thank you for the thorough review, I really appreciate it! 🙏

**Comment by sev-** [comment]
Thank you!

**Pull request merged**
